Errors due to disallow_any_expr not always reported in incremental mode

I found this in an audit of internal mypy runs with and without incremental mode. It seems that when you change the `disallow_any_expr` flag, then incremental mode doesn't consider the cache dirty, so it won't recheck files that might be affected. The same may be true for other `disallow_any*` flags.

However I'm still surprised, because the code in options.py does seem to add these flags to `OPTIONS_AFFECTING_CACHE`.

Here's a repro.

- In file `gid.py`:
```py
from typing import Any

class Gid:
  def __init__(self, arg: Any) -> None: pass
```

- In file `main.py`:
```py
from gid import Gid

def f(a: object) -> None:
    isinstance(a, Gid)
```

- Do a full mypy run:
```
$ mypy --show-column-numbers --disallow-any-expr _.py 
_.py:4:18: error: Expression type contains "Any" (has type "Type[Gid]")
```

- - But an incremental mypy run is clean!
```
$ mypy -i --show-column-numbers --disallow-any-expr _.py
```

Here's the repro for the untyped property -- it's pretty similar.

- File `ns.py`:
```py
class Namespace:
    @property
    def user(self):
        pass
```

- File `main.py`:
```py
import ns
user = ns.Namespace.user
```

- Run `mypy --show-column-numbers --disallow-any-expr main.py`:
```
main.py:2:7: error: Expression type contains "Any" (has type "Callable[[Any], Any]")
```

- Run it with `-i`, `mypy -i --show-column-numbers --disallow-any-expr main.py` gives no errors.

Actually it hasn't got anything to do with untyped properties per se -- I can repro a similar result as in the previous comment with the following in `ns.py`:
```py
class Namespace:
    def __init__(self):
        self.user = f()
def f(): pass
```
This presents as
```
main.py:2:7: error: Expression has type "Any"
```
which is not shown with `-i`.

--- a/mypy/types.py
+++ b/mypy/types.py
@@ -327,12 +327,15 @@ class AnyType(Type):
         return isinstance(other, AnyType)
 
     def serialize(self) -> JsonDict:
-        return {'.class': 'AnyType', 'type_of_any': self.type_of_any.name}
+        return {'.class': 'AnyType', 'type_of_any': self.type_of_any.name,
+                'source_any': self.source_any.serialize() if self.source_any is not None else None}
 
     @classmethod
     def deserialize(cls, data: JsonDict) -> 'AnyType':
         assert data['.class'] == 'AnyType'
-        return AnyType(TypeOfAny[data['type_of_any']])
+        source = data['source_any']
+        return AnyType(TypeOfAny[data['type_of_any']],
+                       AnyType.deserialize(source) if source is not None else None)
 
 
 class UninhabitedType(Type):

--- a/mypy/types.py
+++ b/mypy/types.py
@@ -327,12 +327,12 @@ class AnyType(Type):
         return isinstance(other, AnyType)
 
     def serialize(self) -> JsonDict:
-        return {'.class': 'AnyType'}
+        return {'.class': 'AnyType', 'type_of_any': self.type_of_any.name}
 
     @classmethod
     def deserialize(cls, data: JsonDict) -> 'AnyType':
         assert data['.class'] == 'AnyType'
-        return AnyType(TypeOfAny.special_form)
+        return AnyType(TypeOfAny[data['type_of_any']])
 
 
 class UninhabitedType(Type):

--- a/test-data/unit/check-incremental.test
+++ b/test-data/unit/check-incremental.test
@@ -3447,3 +3447,22 @@ import m  # No error here
 tmp/m/a.py:1: error: Unsupported operand types for + ("int" and "str")
 [out2]
 tmp/m/a.py:1: error: Unsupported operand types for + ("int" and "str")
+
+[case testDisallowAnyExprIncremental]
+# cmd: mypy -m main
+# flags:  --disallow-any-expr
+
+[file ns.py]
+class Namespace:
+    def __init__(self):
+        self.user = 0
+
+[file main.py]
+import ns
+user = ns.Namespace.user
+
+[out1]
+tmp/main.py:2: error: Expression has type "Any"
+
+[out2]
+tmp/main.py:2: error: Expression has type "Any"

